### PR TITLE
Allow Content-Type to be specified and saved as S3 metadata

### LINKED ISSUE #201: Ensure content type and filename information is appropriately passed to storage
Ideally, this would result in MinIO / S3 automatically setting the `Content-Type` and `Content-Disposition` (the filename portion) headers for every download.

This probably involves setting the appropriate S3-Object metadata when the object is first created. The particular headers should be empirically verified. [This may be a starting point for S3 headers](https://github.com/girder/girder/blob/1bfe2849b560fa6b6d25620d6b47dac60ed0cc4b/girder/utility/s3_assetstore_adapter.py#L150-L151).

The original information about the filename and content type should probably come from the client, but the server may also be able to infer content type (see #52).

This was originally reported in #197.

### PR DESCRIPTION
**Pull request by dchiquito**

### CODE REVIEW

File: s3_file_field/_multipart.py
Comment (zachmullen) [inline s3_file_field/_multipart.py:69]: I must be missing something here... when I try this locally in python 3.8 it doesn't appear to be a valid way to make a union type, though I see there's a [PEP for it](https://www.python.org/dev/peps/pep-0604/), but it is not accepted until 3.10. And yet CI is passing. What's going on?

Comment (zachmullen) [reply s3_file_field/_multipart.py:69]: Nevermind, just noticed `from __future__ import annotations`. Carry on.

File: s3_file_field/_multipart_boto3.py
Comment (zachmullen) [inline s3_file_field/_multipart_boto3.py:25]: Let's rename `metadata` to `boto3_kwargs` or something, just because there's actually a separate kwarg to `create_multipart_upload` called `Metadata`, for setting metadata on the object in S3.

File: s3_file_field/_multipart_minio.py
Comment (zachmullen) [inline s3_file_field/_multipart_minio.py:27]: This symbol `_new_multipart_upload` doesn't exist on the latest version of minio-py. It's been replaced with

```python
def _create_multipart_upload(self, bucket_name, object_name, headers)
```

Should we (or is it possible to) upgrade our dependency and use this newer method instead? @brianhelba

File: s3_file_field/views.py
Comment (zachmullen) [inline s3_file_field/views.py:92]: Guessing is OK, but if we want more deterministic / well defined behavior, we could just fall back to `application/octet-stream` as a generic type. If we do that, we can just do

```python
content_type = upload_request.get('content_type', 'application/octet-stream')
```

File: s3_file_field/views.py
Comment (zachmullen) [inline s3_file_field/views.py:90]: `None` is the default for the second argument, so it can be omitted.

File: s3_file_field/views.py
Comment (zachmullen) [inline s3_file_field/views.py:96]:
```suggestion
    content_disposition = upload_request.get('content_disposition, f'attachment; filename="{file_name}"')
```

File: tests/test_views.py
Comment (zachmullen) [inline tests/test_views.py:172]: This test case is getting very complex, which could make it expensive to maintain. Is there a sane way to break it up a bit?

( @brianhelba may have something to say about this )

**Review by zachmullen — CHANGES_REQUESTED** [review]

File: s3_file_field/_multipart.py
Comment (dchiquito) [reply s3_file_field/_multipart.py:69]: I was confused by this as well. I tried to use this exact code in `_multipart_minio.py` and `_multipart_boto3.py` and it broke CI and raised errors in my editor. In this file it seems totally benign.

In lieu of any understanding I can just use `Union` instead.

Comment (zachmullen) [reply s3_file_field/_multipart.py:69]: I defer to other maintainers of this repo as to whether they want this syntax to be allowed. Frankly, I really like it as a way of expressing union types, but it's clearly not yet familiar to mainstream python, and still requires an import side effect, so that might be a turnoff.

File: s3_file_field/views.py
Comment (dchiquito) [reply s3_file_field/views.py]: Not specifying `Content-Type` in the metadata still results in S3 providing `Content-Type: application/octet-stream` when downloading it, so I would think just not specifying would be more sensible than explicitly recording it in metadata.

Comment (zachmullen) [reply s3_file_field/views.py]: Makes sense to me. If it's as simple as passing `ContentType=None` to the boto3 call, we can just do 

```python
content_type = upload_request.get('content_type')
```

File: tests/test_views.py
Comment (dchiquito) [reply tests/test_views.py:172]: Copying the test for each use case would just double the complexity, so I've been avoiding it.

The only solution I can think of would be a test helper that does most of the upload process on it's own, which I consider something of an antipattern. I'm open to suggestions though, this is a very bloated test.

Comment (zachmullen) [reply tests/test_views.py:172]: The body itself is big and branchy, but also just the cardinality of the cartesian product of all these `parametrize` decorators is getting big. That's fine if we really want to test that whole space, but if not, it'd be good to identify the most important cases and maybe collapse the decorators some.

File: s3_file_field/views.py
Comment (dchiquito) [reply s3_file_field/views.py]: Is guessing the type a bad thing? I thought it was a useful sounding feature.

Comment (zachmullen) [reply s3_file_field/views.py]: It's not a bad thing. But I'm also not sure if it's a good thing, and if we are ambivalent, I would always err on the side of the simplest implementation.

File: s3_file_field/views.py
Comment (brianhelba) [inline s3_file_field/views.py:22]: I don't think we should take a `content_disposition ` directly from the client. It should always use stored as `attachment` (and only because we have to make a decision). The client is already providing a `file_name`, so we can just use that.

It's a separate matter for clients to choose to override `Content-Disposition` on an individual download basis, using pre-signed GET request URLs.

File: s3_file_field/views.py
Comment (brianhelba) [inline s3_file_field/views.py:21]: Why not `trim_whitespace`?

File: s3_file_field/_multipart.py
Comment (brianhelba) [reply s3_file_field/_multipart.py:69]: I want us to have nice things, but I think we should move to this syntax as part of a separate PR and code style policy change.

The intent of `from __future__ import annotations` was just to enable delayed evaluation of annotations, which was necessary for...something here; if the code now works without that `__future__` change, I think we should cut it (until we deliberately need it again).

File: s3_file_field/_multipart_minio.py
Comment (brianhelba) [reply s3_file_field/_multipart_minio.py:26]: @zachmullen No, we are pinned to minio-py v6 (the latest is v7), because v7 has a significantly different and incompatible API, which is not (yet) supported by django-minio-storage.

File: s3_file_field/_multipart.py
Comment (brianhelba) [inline s3_file_field/_multipart.py:66]:
```suggestion
        content_type: Optional[str] = None,
```

This is the more canonical syntax.

File: s3_file_field/_multipart.py
Comment (brianhelba) [inline s3_file_field/_multipart.py:155]:
```suggestion
        content_type: Optional[str] = None,
```

File: s3_file_field/_multipart_boto3.py
Comment (brianhelba) [inline s3_file_field/_multipart_boto3.py:30]: What's the MyPy error here? If it can't be fixed, could you use a more specific ignore?

File: s3_file_field/_multipart_boto3.py
Comment (brianhelba) [inline s3_file_field/_multipart_boto3.py:24]: Since we know the file name at this point, can we set `ContentDisposition` too?

File: s3_file_field/_multipart_minio.py
Comment (brianhelba) [inline s3_file_field/_multipart_minio.py:22]: Since we know the file name at this point, can we set `Content-Disposition` too?

File: s3_file_field/_multipart_minio.py
Comment (brianhelba) [inline s3_file_field/_multipart_minio.py:25]: Not clear on what this comment means.

File: tests/test_views.py
Comment (brianhelba) [reply tests/test_views.py]: I agree, I think we should rework this whole test. I'm ok with merging the PR with this, as long as we refactor things at some point.

Comment (zachmullen) [reply tests/test_views.py]: I'm not 100% sure but it looks like maybe @dchiquito already went through and removed some of the complexity here. It looks less scary to me now, at least in terms of this changeset.

File: s3_file_field/_multipart_boto3.py
Comment (dchiquito) [reply s3_file_field/_multipart_boto3.py:24]: We could, but AWS already provides a sensible default value when it is not set. I don't think at this point we would be able to make a better guess than that.

File: s3_file_field/_multipart_boto3.py
Comment (brianhelba) [reply s3_file_field/_multipart_boto3.py]: I have an idea for how to fix this. I'll try locally.

File: s3_file_field/_multipart.py
Comment (brianhelba) [inline s3_file_field/_multipart.py:1]: @dchiquito I just noticed, why is this getting removed? This is what allows the `MultipartManager` type annotation to be forward-referenced without making it a string.

**Review by brianhelba — APPROVED** [review]
One question, otherwise this LGTM!

File: s3_file_field/_multipart.py
Comment (dchiquito) [reply s3_file_field/_multipart.py]: I missed that I had stringly typed `MultipartManager` and thought it was unused. I've restored it.

**Pull request merged**
